src/pages/api/bot-firewall/block-ip.ts

```diff
@@ -0,0 +1,110 @@
+import { syncFirewallRuleset } from '../../../server/botd-firewall/cloudflareApiHelper';
+import { FingerprintJsServerApiClient, isEventError } from '@fingerprintjs/fingerprintjs-pro-server-api';
+import { ALLOWED_REQUEST_TIMESTAMP_DIFF_MS, BACKEND_REGION, SERVER_API_KEY } from '../../../server/const';
+import { ensurePostRequest } from '../../../server/server';
+import { isRequestIdFormatValid, originIsAllowed, visitIpMatchesRequestIp } from '../../../server/checks';
+import { isIP } from 'is-ip';
+import { EventResponseIdentification } from '../../../shared/types';
+
+export type BlockIpPayload = {
+  ip: string;
+  blocked: boolean;
+  requestId: string;
+};
+
+export type BlockIpResponse = {
+  result: 'success' | 'error';
+  message: string;
+  ip?: string;
+  blocked?: boolean;
+};
+
+export default async function blockIp(req: NextApiRequest, res: NextApiResponse<BlockIpResponse>) {
+  // This API route accepts only POST requests.
+  if (!ensurePostRequest(req, res)) {
+    return;
+  }
+
+  // Validate block/unblock request
+  const { ip, blocked, requestId } = req.body as BlockIpPayload;
+  const { okay, message } = await validateBlockIpRequest(requestId, ip, req);
+  if (!okay) {
+    return res.status(403).json({ result: 'error', message });
+  }
+
+  try {
+    // Save or remove blocked IP from database
+    if (blocked) {
+      await saveBlockedIp(ip);
+    } else {
+      await deleteBlockedIp(ip);
+    }
+
+    // Construct updated firewall rules from the blocked IP database and apply them to your Cloudflare application
+    await syncFirewallRuleset();
+
+    // Return success
+    return res.status(200).json({ result: 'success', message: 'OK', ip, blocked });
+  } catch (error) {
+    console.log(error);
+    // Catch unexpected errors and return 500 to the client
+    return res.status(500).json({ result: 'error', message: 'Internal server error.' });
+  }
+}
+
+type ValidationResult = {
+  okay: boolean;
+  message: string;
+};
+
+// For this demo, we want to only enable visitors to block their own IP, not other people's IPs, to not interfere with their demo experience.
+// In the actual scenario, only your authenticated admins could block bot IPs (or they would be blocked automatically) so this check would not be necessary.
+const validateBlockIpRequest = async (
+  requestId: string,
+  ip: string,
+  req: NextApiRequest,
+): Promise<ValidationResult> => {
+  if (!isRequestIdFormatValid(requestId)) {
+    return { okay: false, message: 'Invalid request ID format.' };
+  }
+
+  if (!isIP(ip)) {
+    return { okay: false, message: 'Invalid IP address.' };
+  }
+
+  let identification: EventResponseIdentification;
+  try {
+    const client = new FingerprintJsServerApiClient({ region: BACKEND_REGION, apiKey: SERVER_API_KEY });
+    const eventResponse = await client.getEvent(requestId);
+    identification = eventResponse.products?.identification?.data;
+  } catch (error) {
+    console.log(error);
```

Comment:
```suggestion
    console.error(error);
```

If we need to log it, let's log it as error 😁. Pls also change it in other places as well, or remove the logs.

Reply:
Thanks, I changed all error console.logs to `console.error` :) 
https://github.com/fingerprintjs/fingerprintjs-pro-use-cases/pull/111/commits/f913b5d0da29be1e5ce8d730d13f72415b6ace1c